src/Microsoft.IdentityModel.Validators/AadIssuerValidator/AadIssuerValidator.cs

```diff
@@ -390,14 +389,62 @@ private static bool IsValidIssuer(string validIssuerTemplate, string tenantId, s
 
             if (validIssuerTemplate.Contains(TenantIdTemplate))
             {
-                return validIssuerTemplate.Replace(TenantIdTemplate, tenantId) == actualIssuer;
+                return IssuersWithTemplatesAreEqual(validIssuerTemplate, TenantIdTemplate, actualIssuer, tenantId);
             }
             else
             {
                 return validIssuerTemplate == actualIssuer;
             }
         }
 
+        /// <summary>
+        /// Compare two Issuers with templates without string allocations.
+        /// </summary>
+        internal static bool IssuersWithTemplatesAreEqual(string issuer1, string issuer1Template, string issuer2, string issuer2Template)
+        {
+            int firstHalfIssuer1EndIndex = issuer1.IndexOf(issuer1Template, StringComparison.Ordinal);
+            int firstHalfIssuer2EndIndex = issuer2.IndexOf(issuer2Template, StringComparison.Ordinal);
+
+            // ensure indices exist for both issuers
+            if (firstHalfIssuer1EndIndex == -1  || firstHalfIssuer2EndIndex == -1)
+                return false;
+
+            // if the indices aren't equal, then the issuers aren't equal
+            if (firstHalfIssuer1EndIndex != firstHalfIssuer2EndIndex)
+                return false;
+
+            // ensure the first part of the issuer1 matches the first part of issuer2
+            if (!StringSegmentsAreEqual(issuer1, 0, firstHalfIssuer1EndIndex, issuer2, 0, firstHalfIssuer2EndIndex))
+                return false;
+
+            int secondHalfIssuer1StartIndex = firstHalfIssuer1EndIndex + issuer1Template.Length;
+            int secondHalfIssuer2StartIndex = firstHalfIssuer2EndIndex + issuer2Template.Length;
+
+            // if the indices aren't equal, then the issuers aren't equal
+            if (secondHalfIssuer1StartIndex != secondHalfIssuer2StartIndex)
+                return false;
+
+            // ensure the second halves are equal
+            if (!StringSegmentsAreEqual(issuer1, secondHalfIssuer1StartIndex, issuer1.Length, issuer2, secondHalfIssuer2StartIndex, issuer2.Length))
+                return false;
+
+            return true;
+        }
+
+        /// <summary>
+        /// Compare two string segments for equality without string allocations.
+        /// </summary>
+        private static bool StringSegmentsAreEqual(string str1, int str1Start, int str1End, string str2, int str2Start, int str2End)
```

Comment:
on netcore we should be able to use span based comparisons

https://learn.microsoft.com/en-us/dotnet/api/system.memoryextensions.compareto?view=net-8.0

and we can keep the ordinal comparisons

Reply:
We can use the Span based comparisons on non-netcore as well.

Reply:
Thanks, using spans now :)